pkg/threescale/connectors/backend/local_cache.go

```diff
@@ -0,0 +1,128 @@
+package backend
+
+import (
+	"time"
+
+	"github.com/3scale/3scale-go-client/client"
+	"github.com/orcaman/concurrent-map"
+)
+
+var now = time.Now
+
+// LocalCache is an implementation of Cacheable providing an in-memory cache
+// and asynchronous reporting for 3scale backend
+type LocalCache struct {
+	ds       cmap.ConcurrentMap
+	reporter *reporter
+}
+
+type reporter struct {
+	interval time.Duration
+	stop     chan struct{}
+	cache    cmap.ConcurrentMap
+}
+
+// NewLocalCache returns a pointer to a LocalCache with an initialised empty data structure
+func NewLocalCache(reportInterval *time.Duration, reportChan chan struct{}) *LocalCache {
+	var defaultInterval = time.Second * 15
+
+	if reportInterval == nil {
+		reportInterval = &defaultInterval
+	}
+
+	cache := cmap.New()
+
+	return &LocalCache{
+		ds: cache,
+		reporter: &reporter{
+			interval: *reportInterval,
+			stop:     reportChan,
+			cache:    cache,
+		},
+	}
+}
+
+// Get entries for LocalCache
+func (l LocalCache) Get(cacheKey string) (CacheValue, bool) {
+	var cv CacheValue
+	v, ok := l.ds.Get(cacheKey)
+	if !ok {
+		return cv, ok
+	}
+
+	cv = v.(CacheValue)
+	return cv, ok
+}
+
+// Set entries for LocalCache
+func (l LocalCache) Set(cacheKey string, cv CacheValue) {
+	l.ds.Set(cacheKey, cv)
+}
+
+// Report cached entries to 3scale backend
+func (l LocalCache) Report() {
+	ticker := time.NewTicker(l.reporter.interval)
+	ascendingPeriodSequence := []client.LimitPeriod{client.Minute, client.Hour, client.Day, client.Week, client.Month, client.Eternity}
+	reportMetrics := client.Metrics{}
+	for {
+		select {
```

Comment:
@unleashed - thinking about this function, would it be better to leave it as a blocking function and let the caller worry about running it in the background or do you prefer the approach taken below.